nixos/modules/security/auditd.nix

```diff
@@ -4,15 +4,82 @@
   pkgs,
   ...
 }:
+let
+  cfg = config.security.auditd;
 
+  prepareConfigValue =
+    v:
+    if lib.isBool v then
+      (if v then "yes" else "no")
+    else if lib.isList v then
+      lib.concatStringsSep " " (map prepareConfigValue v)
+    else
+      builtins.toString v;
+  prepareConfigText =
+    conf:
+    lib.concatLines (
+      lib.mapAttrsToList (k: v: if v == null then "#${k} =" else "${k} = ${prepareConfigValue v}") conf
+    );
+in
 {
-  options.security.auditd.enable = lib.mkEnableOption "the Linux Audit daemon";
+  options.security.auditd = {
+    enable = lib.mkEnableOption "the Linux Audit daemon";
 
-  config = lib.mkIf config.security.auditd.enable {
+    config = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType =
+          with lib.types;
+          attrsOf (
+            nullOr (oneOf [
+              bool
+              nonEmptyStr
+              path
+              int
+            ])
+          );
+        options = {
+          # space_left needs to be larger than admin_space_left, yet they default to be the same if left open.
```

Comment:
Should this be an assertion?

Reply:
wrote an assert. It will *not* check the case where one is % and the other absolute, but it does check %/% and abs/abs

Reply:
I have a gut feeling that this is one of those too complex assertions that will become a maintancen burden but since it's already written I won't ask you to undo it.

Reply:
maybe, and then i can still remove it. But the regex to match for definitions is quite strict, so there shouldn't be anything unexpected happening here.